File: source/kernel/malloc.zig
//
// malloc.zig
//
// Copyright (C) 2025 Mateusz Stadnik <matgla@live.com>
//
// This program is free software: you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation, either version
// 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be
// useful, but WITHOUT ANY WARRANTY; without even the implied
// warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
// PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General
// Public License along with this program. If not, see
// <https://www.gnu.org/licenses/>.
//

const std = @import("std");
const c = @cImport({
    @cInclude("stdlib.h");
    @cInclude("string.h");
});

const log = &@import("../log/kernel_log.zig").kernel_log;

const process_memory_pool = @import("process_memory_pool.zig");

const MallocAllocator = struct {
    fn alloc(
        _: *anyopaque,
        len: usize,
        log2_align: std.mem.Alignment,
        return_address: usize,
    ) ?[*]u8 {
        _ = log2_align;
        _ = return_address;
        std.debug.assert(len > 0);
        const ptr = @as([*]u8, @ptrCast(c.malloc(len) orelse return null));

        log.print("malloc alloc: {d}, {*}\n", .{ len, ptr });
        return ptr;
    }

    fn resize(
        ctx: *anyopaque,
        buf: []u8,
        log2_buf_align: std.mem.Alignment,
        new_len: usize,
        return_address: usize,
    ) bool {
        _ = ctx;
        _ = return_address;
        _ = log2_buf_align;
        _ = buf;
        _ = new_len;
        return false;
    }

    fn remap(
        context: *anyopaque,
        memory: []u8,
        alignment: std.mem.Alignment,
        new_len: usize,
        return_address: usize,
    ) ?[*]u8 {
        return if (resize(context, memory, alignment, new_len, return_address)) memory.ptr else null;
    }

    fn free(
        _: *anyopaque,
        buf: []u8,
        log2_buf_align: std.mem.Alignment,
        return_address: usize,
    ) void {
        _ = log2_buf_align;
        _ = return_address;
        log.print("malloc free: {*}:{d}\n", .{ buf.ptr, buf.len });
        c.free(buf.ptr);
    }
};

pub const malloc_allocator = std.mem.Allocator{
    .ptr = undefined,
    .vtable = &malloc_allocator_vtable,
};

const malloc_allocator_vtable = std.mem.Allocator.VTable{
    .alloc = MallocAllocator.alloc,
    .resize = MallocAllocator.resize,
    .free = MallocAllocator.free,
    .remap = MallocAllocator.remap,
};

pub const ProcessPageAllocator = struct {
    pid: u32,

    fn alloc(
        ctx: *anyopaque,
        len: usize,
        log2_align: std.mem.Alignment,
        return_address: usize,
    ) ?[*]u8 {
        const self: *ProcessPageAllocator = @ptrCast(@alignCast(ctx));
        _ = log2_align;
        _ = return_address;
        const number_of_pages: i32 = @intCast((len + process_memory_pool.ProcessMemoryPool.page_size - 1) / process_memory_pool.ProcessMemoryPool.page_size);
        const ptr = @as([*]u8, @ptrCast(process_memory_pool.instance.allocate_pages(number_of_pages, self.pid) orelse return null));
        return ptr;
    }

    fn resize(
        ctx: *anyopaque,
        buf: []u8,
        log2_buf_align: std.mem.Alignment,
        new_len: usize,
        return_address: usize,
    ) bool {
        _ = ctx;
        _ = buf;
        _ = return_address;
        _ = log2_buf_align;
        _ = new_len;

        return false;
    }

    fn remap(
        context: *anyopaque,
        memory: []u8,
        alignment: std.mem.Alignment,
        new_len: usize,
        return_address: usize,
    ) ?[*]u8 {
        return if (resize(context, memory, alignment, new_len, return_address)) memory.ptr else null;
    }

    fn free(
        ctx: *anyopaque,
        buf: []u8,
        log2_buf_align: std.mem.Alignment,
        return_address: usize,
    ) void {
        const self: *ProcessPageAllocator = @ptrCast(@alignCast(ctx));
        _ = log2_buf_align;
        _ = return_address;
        const number_of_pages: i32 = @intCast((buf.len + process_memory_pool.ProcessMemoryPool.page_size - 1) / process_memory_pool.ProcessMemoryPool.page_size);
        process_memory_pool.instance.free_pages(buf.ptr, number_of_pages, self.pid);
    }

    pub fn create(pid: u32) ProcessPageAllocator {
        return ProcessPageAllocator{
            .pid = pid,
        };
    }

    pub fn release_pages(self: *ProcessPageAllocator) void {
        process_memory_pool.instance.release_pages_for(self.pid);
    }

    pub fn std_allocator(self: *ProcessPageAllocator) std.mem.Allocator {
        return std.mem.Allocator{
            .ptr = @ptrCast(self),
            .vtable = &process_page_allocator_vtable,
        };
    }
};

const process_page_allocator_vtable = std.mem.Allocator.VTable{
    .alloc = ProcessPageAllocator.alloc,
    .resize = ProcessPageAllocator.resize,
    .free = ProcessPageAllocator.free,
    .remap = ProcessPageAllocator.remap,
};

File: libs/libc/malloc.c
/*
 *   Copyright (c) 2025 Mateusz Stadnik

 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.

 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.

 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "stddef.h"

#include "stdalign.h"
#include "unistd.h"

typedef struct memory_block {
  size_t size : 31;
  size_t used : 1;
  struct memory_block *next;
} memory_block;

memory_block root = {0, 0, NULL};
memory_block *last = &root;

static memory_block *find_free_block(size_t size) {
  memory_block *current = &root;
  while (current != NULL) {
    if (!current->used && current->size >= size) {
      return current;
    }
    current = current->next;
  }
  return NULL;
}

static size_t malloc_align(size_t size) {
  return (size + alignof(max_align_t) - 1) & ~(alignof(max_align_t) - 1);
}

void *malloc(size_t size) {
  const size_t memory_block_size = malloc_align(sizeof(memory_block));
  const size_t allocation_size = malloc_align(size) + memory_block_size;

  memory_block *free_block = find_free_block(malloc_align(size));
  if (free_block == NULL) {
    void *new_block = sbrk(allocation_size);
    memset(new_block, 0, memory_block_size);
    if (new_block == (void *)-1) {
      return NULL;
    }
    memory_block *block = (memory_block *)new_block;
    block->size = malloc_align(size);
    block->used = 1;
    block->next = NULL;

    last->next = block;
    last = block;
    return new_block + memory_block_size;
  } else {
    if (free_block->size > size + sizeof(memory_block)) {
      memory_block *new_block =
          (memory_block *)((void *)free_block + allocation_size);
      new_block->size = free_block->size - allocation_size;
      new_block->used = 0;
      new_block->next = free_block->next;
      free_block->next = new_block;
      free_block->size = malloc_align(size);
      free_block->used = 1;
      return (void *)free_block + memory_block_size;
    } else {
      free_block->used = 1;
      return (void *)free_block + memory_block_size;
    }
  }

  return NULL;
}

// TODO: consolidate implementation
void free(void *ptr) {
  if (ptr == NULL) {
    return;
  }
  memory_block *block =
      (memory_block *)((void *)ptr - malloc_align(sizeof(memory_block)));

  if (block->used == 0) {
    return;
  }

  memory_block *parent = &root;
  for (parent = &root; parent != NULL; parent = parent->next) {
    if (parent->next == block) {
      break;
    }
  }
  block->used = 0;
  if (parent != NULL) {
    if (block->next != NULL && block->next->used == 0) {
      block->size += block->next->size + malloc_align(sizeof(memory_block));
      block->next = block->next->next;
    }
    if (parent->used == 0 && parent != &root) {
      parent->size += block->size + malloc_align(sizeof(memory_block));
      parent->next = block->next;
      if (last == block) {
        last = parent;
      }
      block = parent;
    }
  }

  if (last == block) {
    sbrk(-(block->size + malloc_align(sizeof(memory_block))));
    last = parent;
    if (parent != NULL) {
      parent->next = block->next;
    }
    return;
  }
}